gobblin-yarn/src/main/java/org/apache/gobblin/yarn/HelixInstancePurgerWithMetrics.java

```diff
@@ -0,0 +1,140 @@
+            + "non-thread safe call. To disable purging offline instances during startup, change the flag {} "
+            + "elapsedTimeMs={}, laggingThresholdMs={}",
+            GobblinYarnConfigurationKeys.HELIX_PURGE_OFFLINE_INSTANCES_ENABLED, elapsedTimeMs, laggingThresholdMs);
+        if (!haveSubmittedLaggingEvent && elapsedTimeMs > laggingThresholdMs) {
+          submitLaggingEvent(elapsedTimeMs, laggingThresholdMs, gteMetadata);
+          haveSubmittedLaggingEvent = true;
+        }
+        Thread.sleep(this.pollingRateMs);
+      }
+
+      long timeToPurgeMs = watch.elapsed(TimeUnit.MILLISECONDS);
+      if (!haveSubmittedLaggingEvent && timeToPurgeMs > laggingThresholdMs) {
+        submitLaggingEvent(timeToPurgeMs, laggingThresholdMs, gteMetadata);
+      }
+
+      purgeTask.get(); // check for exceptions
+      submitCompletedEvent(timeToPurgeMs, gteMetadata);
+      return timeToPurgeMs;
+    } catch (ExecutionException | InterruptedException e) {
+      log.info("The call to purge offline helix instances failed. This is not a fatal error because it is not mandatory to "
+          + "clean up old helix instances. But repeated failure to purge offline helix instances will cause an accumulation"
+          + "of offline helix instances which may cause large delays in future helix calls.", e);
+      long timeToPurgeMs = watch.elapsed(TimeUnit.MILLISECONDS);
+      submitFailureEvent(timeToPurgeMs, gteMetadata);
+      return timeToPurgeMs;
+    }
+  }
+
+  private void submitFailureEvent(long elapsedTimeMs, Map<String, String> additionalMetadata) {
+    if (eventSubmitter != null) {
+      GobblinEventBuilder eventBuilder = new GobblinEventBuilder(PURGE_FAILURE_EVENT);
+      eventBuilder.addAdditionalMetadata(additionalMetadata);
+      eventBuilder.addMetadata("elapsedTimeMs", String.valueOf(elapsedTimeMs));
+
+      log.info("Submitting GTE because purging offline instances has failed to complete. event={}", eventBuilder);
+      eventSubmitter.submit(eventBuilder);
+    } else {
+      log.info("Cannot submit {} GTE because eventSubmitter is null", PURGE_FAILURE_EVENT);
+    }
+  }
+
+  private void submitCompletedEvent(long timeToPurgeMs, Map<String, String> additionalMetadata) {
+    if (eventSubmitter != null) {
+      GobblinEventBuilder eventBuilder = new GobblinEventBuilder(PURGE_COMPLETED_EVENT);
+      eventBuilder.addAdditionalMetadata(additionalMetadata);
+      eventBuilder.addMetadata("timeToPurgeMs", String.valueOf(timeToPurgeMs));
+
+      log.info("Submitting GTE because purging offline instances has completed successfully. event={}", eventBuilder);
+      eventSubmitter.submit(eventBuilder);
+    } else {
+      log.info("Cannot submit {} GTE because eventSubmitter is null", PURGE_COMPLETED_EVENT);
+    }
+  }
+
+  private void submitLaggingEvent(long elapsedTimeMs, long laggingThresholdMs,
```

Comment:
how do we track these events? are they processed somewhere or do we have any automatic monitoring? or will we query them when there's an issue we're noticing with helix?

Reply:
The events being submitted are Gobblin Tracking Events sent to Kafka. One way to setup general reporting is to query the Hive table specifically for this GobblingTrackingEvent topic. 

It is up to us, the users, to setup the infrastructure to consume these events. 

GobblinTrackingEvents aren't new, so existing monitoring services can easily consume this and emit alerts.